Addition and changes in account.loan module only.

S0ME1ONE opened the pull request:
Major Addition:
- Add account.loan.report module and related view.
- Add Cron job for posting loan lines.
- Add logging message for loan posting, loan lines posting and loan closing.
- Add Arabic translation.

Minor Addition:
- Add description field in account.loan module.
- Add filter domain for deprecated accounts on short_term_loan_account_id, long_term_loan_account_id and interest_expenses_account_id fields.

S0ME1ONE committed:
Addition and changes in account.loan module only.

Major Addition:
- Add account.loan.report module and related view.
- Add Cron job for posting loan lines.
- Add logging message for loan posting, loan lines posting and loan closing.
- Add Arabic translation.
Minor Addition:
- Add description field in account.loan module.
- Add filter domain for deprecated accounts on short_term_loan_account_id, long_term_loan_account_id and interest_expenses_account_id fields.

JordiBForgeFlow commented:
Can you please attend to the Travis errors?

JordiBForgeFlow commented on account_loan/model/account_loan.py:
```diff
@@ -86,7 +86,6 @@ def _default_company(self):
         default=0.0,
         digits=(8, 6),
         help='Currently applied rate',
-        track_visibility='always',
```
why do you remove? This has been put for a reason, and is that once the loan has been confirmed one can change the rate, and we better track the changes.

S0ME1ONE replied:
Thank You @jbeficent for your notes, sorry about that but I have add full logging messages as you can see and I didn't think that rate field was changeable after posting, therefore this change would be reverted now .

JordiBForgeFlow commented on account_loan/report/account_loan_report.py:
```diff
@@ -0,0 +1,71 @@
+    interest_expenses_account_id = fields.Many2one('account.account', string='Interests account', readonly=True)
+
+    @api.model_cr
+    def init(self):
```
Better use:
    @api.model_cr
    def init(self):
        tools.drop_view_if_exists(self.env.cr, self._table)
        self.env.cr.execute("""CREATE or REPLACE VIEW %s as (
            %s
            FROM ( %s )
            %s
            )""" % (self._table, self._select(), self._from(), self._group_by()))

See for example https://github.com/odoo/odoo/blob/10.0/addons/project/report/project_report.py#L89

S0ME1ONE replied:
Thanks again @jbeficent but I don't think we should split one function (method) into three when it's not need and it's sql statement I thik its better stay this way, but I Can make these changes if you insist or we have an issue with this implementation.

JordiBForgeFlow replied:
@imahmedyousif You do it for extensibility purposes. It's a best practice in Odoo. That way one can create modules like account_loan_xxxx that adds fields to the loan and can easily extend the report by calling super to the _select(), _from(),.. methods. Otherwise the sql code is mononlythic and imposible to extend.

S0ME1ONE replied:
Thanks for explaining, I have made requested  the changes.

JordiBForgeFlow reviewed (changes requested):
I made some comments. Please fix travis and we can continue the review.

S0ME1ONE committed:
fix errors and making some requested changesed changes

- Fixing Travis errors.
- Add security access for account.loan.report module.
- Reimplement tracking for rate field.

S0ME1ONE committed:
Fixing Travis test_flake8  error

S0ME1ONE committed:
Implementing the suggested report sql changes

S0ME1ONE committed:
Fixing Travis error

S0ME1ONE committed:
fix test_flake8  error

etobella commented on account_loan/model/account_loan_line.py:
```diff
@@ -307,6 +311,8 @@ def generate_move(self):
                 move = self.env['account.move'].create(record.move_vals())
                 move.post()
                 res.append(move.id)
+        if self.loan_id.state == 'posted':
```
I think you should raise a Warning if the loan is not posted before you try to create the move. Moreover, it should be `if rec.loan_id.state != 'posted':` as you could try to create moves from different loans.

S0ME1ONE replied:
My add code doesn't create moves it just post a log message for the created moves and after digging and searching for a reason to use the ``` for record in self: ``` loop since all function that call generate_move and generate_invoice gives single record not set of record, I found that generate entries module is giving a set of records and that problem could be solved by adding single loop in the mentioned module instead of writing the loop for each related called function and that will shrink the code and free more resources, anyway I shall include the code that highlighted above to the loop in the next commit to avoid singleton error, also no need for the warning since a warning will be raised if the line wasn't posted and log message will be posted only if the move was created and the loan was posted.
the code will be:
``` python
       for record in self:
            if not record.move_ids:
                if record.loan_id.line_ids.filtered(
                        lambda r: r.date < record.date and not r.move_ids
                ):
                    raise UserError(_('Some moves must be created first'))
                move = self.env['account.move'].create(record.move_vals())
                move.post()
                res.append(move.id)
            if record.move_ids and record.loan_id.state == 'posted':
                record.log_posted_line()
```
thank you for your note.

etobella replied:
But if you execute the same function twice for the same line, you will post the text twice, so, if is no necessary, and the log_posted_line should be called before move.post.

S0ME1ONE replied:
In which case that will happen, this function should execute once for the same line and even if that happend, the message will post only if the move line is created for that loan line and I used ` if record.loan_id.state == 'posted' ` just because the loan close message will appear before the line post message if i remove this part, but if you have some thing that I can't see about why this function will be called twice please give me a clue.

>the log_posted_line should be called before move.post.

if I call log_posted_line before, the message will be posted even if the moves didn't created therefore I need to check for move id before posting the message.

etobella commented on account_loan/model/account_loan_line.py:
```diff
@@ -328,6 +334,8 @@ def generate_invoice(self):
                 for line in invoice.invoice_line_ids:
                     line._set_taxes()
                 invoice.compute_taxes()
+        if self.loan_id.state == 'posted':
```
The same as before

S0ME1ONE replied:
Same as above

etobella commented on account_loan/model/account_loan_line.py:
```diff
@@ -382,3 +390,34 @@ def view_account_invoices(self):
+        ])
+        for line in lines:
+            if line.is_leasing:
+                line.generate_invoice()
```
You could do `line.filtered(lambda r: r.is_leasing).generate_invoice()` and `line.filtered(lambda r: not r.is_leasing).generate_move()` as functions are available for multiple lines.

S0ME1ONE replied:
Thank you @etobella for your comments.
- About version: should we consider this a major update and what the ver. will be? (11.1.0 or 11.0.2).
- About translation: honestly I don't have enough information about OCA and transifex but I like to include the translation of my language for offline use purposes.

S0ME1ONE replied:
Yes, I could do that but I think the used code is simpler and there is no difference in result.
Thank you for your suggestion.

etobella replied:
Well, you are using two fors when you could use one.

etobella replied:
About translation, I would prefer other people opinions. However, when modules are updated, a good practice is to squash translation commits, but if a translation is made in a regular commit it always gives problems. So, I think you should make the translation in transifex. You can access transifex with these link:
https://www.transifex.com/oca/OCA-account-financial-tools-11-0/

S0ME1ONE replied:
I don't see two "for" loops in the highlighted code, but if you mean the "for" loop in _cron_generate_loan_entries function, I decided to leave it incase someone fix the issue we discussed before and it will not hurt right now since only one called function has another useless loop (which is not mine).

S0ME1ONE replied:
@jbeficent @etobella 
I will revert the changes for translation and try to work it with Transifex.

etobella reviewed (changes requested):
You should modify version also.
Just a question, translation should be done on this kind of commits or leave them for transifex?
It's a question I had several times and I don't know what is the general opinion in oca projects. If you leave it in usual commits is complicated to squash OCA Transbot commits on migrations.

S0ME1ONE committed:
Fixing error and add

- Fixing singleton error when generating entries from "generate entries" module.
- Adding logging message for leasing loan.

JordiBForgeFlow commented:
You can propose your own translation manually, in a specific commit, or do
it in Transifex. If you do it manually, much better to clearly indicate
'Translation to Xxxx' in the commit description.

Whoever tries to squash tranlation commits will take it into account.

Guide on squashing translation commits:
https://github.com/OCA/maintainer-tools/wiki/Merge-commits-in-pull-requests#mergesquash-the-commits-generated-by-transifex

On Sun, Apr 15, 2018, 7:51 AM Enric Tobella <notifications@github.com>
wrote:

> *@etobella* commented on this pull request.
> ------------------------------
>
> In account_loan/model/account_loan_line.py
> <https://github.com/OCA/account-financial-tools/pull/647#discussion_r181570989>
> :
>
> > @@ -382,3 +390,34 @@ def view_account_invoices(self):
>              result['views'] = [(res and res.id or False, 'form')]
>              result['res_id'] = self.invoice_ids.id
>          return result
> +
> +    @api.model
> +    def _cron_generate_loan_entries(self):
> +        date_now = datetime.now().strftime(DEFAULT_SERVER_DATE_FORMAT)
> +        lines = self.search([
> +            ('has_moves', '=', False),
> +            ('has_invoices', '=', False),
> +            ('loan_id.state', '=', 'posted'),
> +            ('date', '<=', date_now),
> +        ])
> +        for line in lines:
> +            if line.is_leasing:
> +                line.generate_invoice()
>
> About translation, I would prefer other people opinions. However, when
> modules are updated, a good practice is to squash translation commits, but
> if a translation is made in a regular commit it always gives problems. So,
> I think you should make the translation in transifex. You can access
> transifex with these link:
> https://www.transifex.com/oca/OCA-account-financial-tools-11-0/
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/OCA/account-financial-tools/pull/647#discussion_r181570989>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHU_Vp2kvfO6QjqNf9kUIEH6M9EJ7Ezmks5tot_6gaJpZM4TSzmS>
> .
>

S0ME1ONE committed:
Reverting translation changes
